src/test/components/elements/panels/ManageExistingBookings.test.tsx

```diff
@@ -0,0 +1,183 @@
+  tableHeader.forEach((header, index) => expect(header).toHaveTextContent(expectedHeaders[index]));
+};
+
+const getTableRowCells = (row: HTMLElement) => within(row).getAllByRole("cell");
+
+const expectRowData = (row: HTMLElement, expectedData: string[]) => {
+  const rowCells = getTableRowCells(row);
+  expect(rowCells).toHaveLength(expectedData.length + 1);
+  rowCells.slice(1).forEach((cell, index) => expect(cell).toHaveTextContent(expectedData[index]));
+};
+
+const getActionButton = async (firstBookingRow: HTMLElement, buttonName: string) => {
+  const firstBookingCells = within(firstBookingRow).getAllByRole("cell");
+  const firstBookingActionsCell = firstBookingCells[0];
+  return within(firstBookingActionsCell).queryByRole("button", { name: new RegExp(buttonName, "i") });
+};
+
+describe("ManageExistingBookings", () => {
+  const setupTest = (eventBookings: EventBookingDTO[], role: UserRole = "ADMIN") => {
+    renderTestEnvironment({
+      role: role,
+      PageComponent: ManageExistingBookings,
+      componentProps: {
+        user: { ...mockUser, loggedIn: true, role: role },
+        eventBookingId: mockEvent.id,
+      },
+      initialRouteEntries: ["/admin/events/"],
+      extraEndpoints: [
+        rest.get(API_PATH + `/events/${mockEvent.id}/bookings`, (req, res, ctx) => {
+          return res(ctx.status(200), ctx.json(eventBookings));
+        }),
+        rest.get(API_PATH + `/users/school_lookup`, (req, res, ctx) => {
+          const mockSchoolData: UserSchoolLookup = {
+            ["200"]: {
+              urn: "123456",
+              name: "Mock School",
+              postcode: "ABC 123",
+              closed: false,
+              dataSource: "GOVERNMENT_UK",
+            },
+          };
+          return res(ctx.status(200), ctx.json(mockSchoolData));
+        }),
+      ],
+    });
+  };
+
+  it("renders table of bookings with correct headers, if bookings are found for an event", async () => {
+    setupTest(mockEventBookingsForMockEvent);
+    expect(screen.getByText("Manage current bookings")).toBeInTheDocument();
+    const { bookingsTable } = await getTableAndFirstRow();
+    const expectedColumnHeaders = [
+      "Actions",
+      "Name",
+      "Email",
+      "Account type",
+      "School",
+      "Job / Year Group",
+      "Booking status",
+      "Gender",
+      "Booking created",
+      "Booking updated",
+      "Stage",
+      "Exam board",
+      "Reserved by ID",
+      "Accessibility requirements",
+      "Medical / Dietary",
+      "Emergency name",
+      "Emergency telephone",
+    ];
+    expectTableHeaders(bookingsTable, expectedColumnHeaders);
+  });
+
+  it("populates expected data in first table row, if bookings are found for an event", async () => {
+    setupTest(mockEventBookingsForMockEvent);
+    const { firstBookingRow } = await getTableAndFirstRow();
+    const firstUser = mockEventBookingsForMockEvent[0].userBooked!;
+    const expectedFirstBookingData = [
```

Comment:
There is too much computed data in the elements of this array. This should be either straight strings, or alternatively you can use variables but get rid of all the logic. For instance:

```typescript
firstUser.email || ""
```

You already know what the first user email is, so that line should be `"my-email@example.com"` or `firstUser.email`.

The way it's done at the moment is not appropriate for a test, this looks more like implementation code. Tests should be easy to read and ideally with no logic at all (to make them easier to read and less prone to errors).

Reply:
I started with hardcoded data, but that caused some issues with dates at least (it passed locally, but not once pushed). Then I changed to the `firstUser.email` type approach, but then the test objects because technically it can be undefined.  Perhaps I should use a mixed approach here and use a calculation on the dates but regular strings for what I know the other values are?

Reply:
Regular strings for the dates should also work... Do you have the error message?

Reply:
It's because the time on server is 1h different to the local time for me - it's something that came up in another test historically when I tried to leave a plain string for a date, so expected vs actual is 1h out when it runs in the pipeline.

Reply:
Oh, that isn't good... Do we have a tech debt ticket to tackle this issue? If so, we should bring it to the next sprint.

I'll approve this PR as it, but if this is localised in this test file, it may be worth looking into refactoring this file to use Jest fake timers, just in case there is an easy solution for it. But if you don't manage to make it work easily, feel fre to merge these changes.

Reply:
Had some difficulty with this today, so opting to create a backlog ticket and address in the new year